packages/react-client/src/hooks/internal/device/useDevices.ts

```diff
@@ -0,0 +1,144 @@
+import { useCallback, useRef, useState } from "react";
+
+import { prepareConstraints } from "../../../devices/constraints";
+import { correctDevicesOnSafari, getAvailableMedia } from "../../../devices/mediaInitializer";
+import type { AudioVideo } from "../../../types/internal";
+import type { DeviceError, PersistLastDeviceHandlers } from "../../../types/public";
+import { useDeviceManager } from "./useDevice";
+import { useHandleTrackEnd } from "./useHandleStreamEnd";
+
+interface UseDevicesProps {
+  videoConstraints?: MediaTrackConstraints | boolean;
+  audioConstraints?: MediaTrackConstraints | boolean;
+  persistHandlers?: PersistLastDeviceHandlers;
+}
+
+type InitializeDevicesResult = { stream: MediaStream | null; errors: AudioVideo<DeviceError | null> | null };
+
+export const useMediaDevices = ({ videoConstraints, audioConstraints, persistHandlers }: UseDevicesProps) => {
+  const [deviceList, setDeviceList] = useState<MediaDeviceInfo[]>([]);
+
+  const [videoStream, setVideoStream] = useState<MediaStream | null>(null);
+  const [audioStream, setAudioStream] = useState<MediaStream | null>(null);
+
+  const [videoError, setVideoError] = useState<DeviceError | null>(null);
+  const [audioError, setAudioError] = useState<DeviceError | null>(null);
+
+  const lastUsedCameraRef = useRef<MediaDeviceInfo | null>(persistHandlers?.getLastDevice("video") ?? null);
+  const lastUsedMicRef = useRef<MediaDeviceInfo | null>(persistHandlers?.getLastDevice("audio") ?? null);
+
+  const initializationRef = useRef<Promise<InitializeDevicesResult> | null>(null);
+
+  const saveUsedCamera = useCallback(
+    (deviceInfo: MediaDeviceInfo) => {
+      lastUsedCameraRef.current = deviceInfo;
+      persistHandlers?.saveLastDevice(deviceInfo, "video");
+    },
+    [persistHandlers],
+  );
+
+  const saveUsedMic = useCallback(
+    (deviceInfo: MediaDeviceInfo) => {
+      lastUsedMicRef.current = deviceInfo;
+      persistHandlers?.saveLastDevice(deviceInfo, "audio");
+    },
+    [persistHandlers],
+  );
+
+  const initializeDevices = useCallback(
+    async (settings?: { enableVideo?: boolean; enableAudio?: boolean }) => {
+      if (deviceList.length) {
+        return null;
```

Comment:
Should we have different return value for cases `there are no devices` and `all is ok`?

Reply:
I introduced a result object with a `status` property that describes the outcome.